### PR TITLE
Ensure assemblies listed `FileList` are not loaded

### LINKED ISSUE #12821: Files listed in `FileList` in module manifest are loaded on module load
Since version `6.2.0-preview.2` all the files listed in `FileList` are loaded when the module is loaded.

This is change in behavior caused by PR #8073 which was done to support a customer issue about loading assemblies from module folder first, before looking at GAC.

Hence, the behavior in `6.2.x` and `7.0.x` is different from Windows PowerShell. 

I believe the current behavior is wrong and should be fixed. But, since it as release in `6.2.x` and `7.0.x` it will be a breaking change. 

If we decide to not fix it, it might make moving from Windows PowerShell to PowerShell 7 more difficult.

I would like opinion from committee about fixing the issue in `7.1` and whether it should be back ported to `7.0.x` and `6.2.x`

# Steps to reproduce

Load a module with managed assemblies in FileList in the psd1

```powershell
$filesBefore = [System.AppDomain]::CurrentDomain.GetAssemblies()
Import-Module <>
$filesAfter = [System.AppDomain]::CurrentDomain.GetAssemblies()
 Compare-Object -ReferenceObject $filesBefore -DifferenceObject $filesAfter
```

# Expected behavior

```none
Files in FileList should not be loaded
```

# Actual behavior

```none
Files in FileList are loaded.
```

# Environment data

```none
Name                           Value
----                           -----
PSVersion                      7.0.1
PSEdition                      Core
GitCommitId                    7.0.1
OS                             Microsoft Windows 10.0.19635
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### PR DESCRIPTION
**Pull request by adityapatwardhan**
Fixes #12821 

# PR Summary

Assemblies listed in `FileList` of the `psd1` should not be loaded when the module is loaded. This change fixes the regression introduced in `6.2.0-preview.2`

## PR Context

Regression was introduced as a side effect of https://github.com/PowerShell/PowerShell/pull/8073

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs [hunk 1]
```diff
@@ -4585,7 +4586,7 @@ internal string FixupFileName(string moduleBase, string name, string extension,
             // Return the path if successfully resolved.
             if (resolvedPath != null)
             {
-                if (isImportingModule && resolvedPath.EndsWith(".dll", StringComparison.OrdinalIgnoreCase))
+                if (isImportingModule && resolvedPath.EndsWith(".dll", StringComparison.OrdinalIgnoreCase) && !skipLoading)
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs:4589]: Why do we need `skipLoading` if we have `isImportingModule`?

File: test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1 [hunk 2]
```diff
@@ -361,3 +361,32 @@ Describe "Import-Module -Force behaviour" -Tag "CI" {
...
+            }
+"@
+
+        $originalPSModulePath = $env:PSModulePath
+        New-Item -ItemType Directory -Path "$testdrive\Modules\FileListLoadTest\" -Force > $null
+
+        $asmPath = "$TESTDRIVE\Modules\FileListLoadTest\FileListLoadTest.dll"
+        Add-Type -TypeDefinition $src -OutputAssembly $asmPath
+
+        $env:PSModulePath += [System.IO.Path]::PathSeparator + "$testdrive\Modules"
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1:383]: We should restore original value in `AfterAll`.

File: src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs (hunk 1 above)
Comment (adityapatwardhan) [reply src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs:4589]: Even though we are importing the module, we should not be loading the files in FileList here. FileList is meant to be just a list of all the files in the module. It should not have effect on the loading.

File: test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1 (hunk 2 above)
Comment (adityapatwardhan) [reply test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1:383]: Fixed.

**Review by iSazonov — APPROVED** [review]
LGTM with one minor comment.

File: src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs (hunk 1 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs:4589]: The new parameter description and name could be more informative.
`getFileListOnly`?

**Comment by adityapatwardhan** [comment]
@adityapatwardhan make sure this repros in 7.0.x

**Comment by TravisEz13** [comment]
@SteveL-MSFT  Do we want to back port this?

**Pull request merged**